Expand stdio adapter's functionality.

geeknoid opened the pull request:
- You can now output to files in addition to stdout and stderr.

- You can now enable automatic file rotation. This works just like
pkg/log.

- You can now constrain the output to a particular log level.

- Outputting in console-friendly mode now actually works. JSON output was always
previously forced even when config said not to.

- Added reasonable defaults for severity level mappings for log entries so hopefully
most operators won't need to specify that.

This has 100% test coverage. An outstanding issue is what to name this adapter now that it does more than stdio. Should I rename it to "locallogging" or something? This would break existing configs. Do we need an aliasing scheme so stdio can be automatically handled by locallogging?

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | b5234b31a4ed6ebbbd49cf41bf3968edd35a8257 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2530/istio-presubmit/2328/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@geeknoid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 4dda379a32f1f5e33529aeb35156876caef32980 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2530/istio-presubmit/2334/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

guptasu commented on mixer/adapter/stdio/config/config.proto:
```diff
@@ -43,12 +53,48 @@ message Params {
     map<string, Level> severity_levels = 2;
 
     // The level to assign to metrics being output.
     Level metric_level = 3;
```
Default ?

guptasu commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -157,7 +165,20 @@ type builder struct {
+			ce = ce.Appendf("outputPath", "cannot specify an output path when using a STDOUT or STDERR log stream")
+		}
+	} else {
+		if b.adapterConfig.OutputPath == "" {
```
Do we want to actually do valid path validation  ?

guptasu commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -157,7 +165,20 @@ type builder struct {
+		}
+	} else {
+		if b.adapterConfig.OutputPath == "" {
+			ce = ce.Appendf("outputPath", "need a valid output path when using a FILE or ROTATING_FILE log stream")
```
ROTATED_FILE

guptasu commented on mixer/adapter/stdio/zap.go:
```diff
@@ -0,0 +1,130 @@
+
+	case config.FILE:
+		outputSink, closer, err = zap.Open(options.OutputPath)
+		if err != nil {
```
can be merged with line above

guptasu commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -139,9 +131,25 @@ func GetInfo() adapter.Info {
+			MetricLevel:        config.INFO,
+			OutputLevel:        config.INFO,
+			OutputAsJson:       true,
+			RotationMaxAge:     30,
```
Shall we rename these to MaxRotationDays. Trying to embed the format in the anme

guptasu replied:
Or RotationDaysMax

guptasu commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -139,9 +131,25 @@ func GetInfo() adapter.Info {
+			OutputLevel:        config.INFO,
+			OutputAsJson:       true,
+			RotationMaxAge:     30,
+			RotationMaxSize:    100 * 1024 * 1024,
```
MaxRotationBytes

guptasu replied:
or RotationBytesMax ?

guptasu commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -139,9 +131,25 @@ func GetInfo() adapter.Info {
+			OutputAsJson:       true,
+			RotationMaxAge:     30,
+			RotationMaxSize:    100 * 1024 * 1024,
+			RotationMaxBackups: 1000,
```
MaxRotationFileCount

guptasu replied:
or RotationFileCountMax

geeknoid commented:
Addressed all feedback, except the bit about validating file paths. Golang doesn't provide a standard path validating function, and I don't want to deal with UNIX vs. Windows stuff in this code, so I think we can skip that. Malformed paths will surface as log errors output to stderr by zap.

guptasu commented:
For file path, will `os.Stat` work ?

geeknoid commented:
No, we don't want. Configs have to validate statically, independently of
the system they are running on. We shouldn't validate that the path exist,
merely that it is well-formed. And since there's no API to check for
well-formedness, well...

On Wed, Jan 10, 2018 at 3:03 PM, Sunny Gupta <notifications@github.com>
wrote:

> For file path, will os.Stat work ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/2530#issuecomment-356767264>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AVucHcygBhHj0QNdoZgwlN___HDg-Rsfks5tJUHNgaJpZM4RZycy>
> .
>

douglas-reid commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -62,11 +64,11 @@ func (h *handler) HandleLogEntry(_ context.Context, instances []*logentry.Instan
 		for _, varName := range h.logEntryVars[instance.Name] {
 			if value, ok := instance.Variables[varName]; ok {
-				fields = append(fields, field(varName, value))
+				fields = append(fields, zap.Any(varName, value))
```
curious: have the IP_ADDRESS issues been resolved?

geeknoid replied:
Yep, we pass the right thing to adapters now. Verified by my tests.

douglas-reid commented on mixer/adapter/stdio/stdio.go:
```diff
@@ -139,9 +131,29 @@ func GetInfo() adapter.Info {
+			OutputLevel:                config.INFO,
+			OutputAsJson:               true,
+			MaxDaysBeforeRotation:      30,
+			MaxMegabytesBeforeRotation: 100 * 1024 * 1024,
```
aside: it would be really nice if there was a size equivalent of `time.Seconds`.

douglas-reid commented on mixer/adapter/stdio/zap.go:
```diff
@@ -0,0 +1,129 @@
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"os"
```
this should be a fmt / lint error. "os" should be in the standard block.

geeknoid replied:
I thought this was normally a lint error, but apparently not.

Fixed now anyway.

douglas-reid commented on mixer/adapter/stdio/zap.go:
```diff
@@ -0,0 +1,129 @@
+	return l, closer, nil
+}
+
+func formatDate(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
```
this looks familiar. possibility for reuse ?

geeknoid replied:
Yes, this was stolen from pkg/log. Do you think it's worth exporting the function from pkg/log? It doesn't really belong in the public surface of pkg/log though. And I don't think it warrants a whole new date formatting package...

geeknoid commented:
/retest all

geeknoid commented:
LGTM someone?

kyessenov commented:
/lgtm
